### PR TITLE
FIX: Conditional Protobuf

### PR DESCRIPTION
**Pull request by tornadia**
## Description
As a Linux User, I don't want to install Google Libraries that I know nothing about in order to compile a client software that I have been using for years without an issue, specially due to the fact that I will not decide to upgrade to 12.70 since:
- I don't have the need yet
- It doesn't have widespread support yet

## Proposed solution
I would like to keep using 7.4-10.98 with the current data structures and conditionally disable protobuf for compilation when using cmake by proportioning the **USE_PROTOBUF** flag.

## Additional info
Please let me know if I should correct anything in this branch

### CODE REVIEW

File: src/framework/CMakeLists.txt [hunk 1]
```diff
@@ -217,57 +217,65 @@ message(STATUS "LuaJIT: " ${LUAJIT})
...
-    ${PROTOBUF_LIBRARY}
-    ${LIBLZMA_LIBRARIES}
-    ${NLOHMANN_JSON_LIBRARY}
 )
 
 set(framework_INCLUDE_DIRS ${framework_INCLUDE_DIRS}
     ${Boost_INCLUDE_DIRS}
     ${LUA_INCLUDE_DIR}
     ${PHYSFS_INCLUDE_DIR}
-    ${framework_INCLUDE_DIRS}
```
Comment (tornadia) [inline src/framework/CMakeLists.txt:238]: this hjere was redundant

**Comment by tornadia** [comment]
Pending: @nekiro or somebody else test if the `.vcxproj` is working on **windows**

File: CMakeLists.txt [hunk 2]
```diff
@@ -18,6 +18,7 @@ if(NOT APPLE)
 endif()
 
 option(USE_PCH "Use precompiled header (speed up compile)" OFF)
+option(USE_PROTOBUF "Use Protobuf for 12.70 support" ON)
```
Comment (nekiro) [inline CMakeLists.txt]: "Use Protobuf for protocol 12+"

File: src/framework/CMakeLists.txt [hunk 3]
```diff
@@ -217,57 +217,67 @@ message(STATUS "LuaJIT: " ${LUAJIT})
...
+                OUTPUT ${protobuf_SOURCES}
+                DEPENDS ${protobuf_PATH}/${protobuf_FILENAME}.proto
+                COMMAND protoc --cpp_out=../ ${protobuf_FILENAME}.proto
+                WORKING_DIRECTORY ${protobuf_PATH}
+                COMMENT "Compiling ${protobuf_FILENAME}.proto"
+            )
+        endif()
+
+    endforeach()
+    set(framework_DEFINITIONS ${framework_DEFINITIONS} -DUSE_PROTOBUF)
```
Comment (oarsc) [inline src/framework/CMakeLists.txt]: You moved all protobuf files to `client` source folder (not sure why) [bbe4f8b](https://github.com/mehah/otclient/pull/238/commits/bbe4f8b18e1235213f1728d999dffe7ef9f34670), but this snippet is trying to compile them from the framework folder...
```
make[2]: *** No rule to make target '../src/framework/protobuf/proto/shared.proto', needed by '../src/framework/protobuf/shared.pb.cc'.  Stop.
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/otclient.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

**Review by oarsc — CHANGES_REQUESTED** [review]

**Comment by nekiro** [comment]
Just to clarify, this doesn't actually work on windows, including conditionally in vcxproj is messy

File: src/framework/CMakeLists.txt (hunk 3 above)
Comment (tornadia) [reply src/framework/CMakeLists.txt]: Aight will fix this

**Comment by tornadia** [comment]
> Just to clarify, this doesn't actually work on windows, including conditionally in vcxproj is messy

It actually those, @nekiro : just need to make a .cpp/.h file to act as an API between the two and then include it in a PCH #if block instead.

**Comment by oarsc** [comment]
Sorry @tornadia I've been a bit absent lately.

I wanted to do a clean test, but the first time I executed `cmake .. -USE_PROTOBUF`, protobuf didn't activate. When I did a second time it worked.
Also, I've found a linking error when it finished. I'm attaching the complete log in a file: [complete-log.txt](https://github.com/mehah/otclient/files/8311369/complete-log.txt)


**Example building the first time, and protobuf doesn't activate**
```sh
scopp@:~/pro/git/otclient/build $ rm ../src/client/protobuf/*.pb.*
scopp@:~/pro/git/otclient/build $ rm -fr *
scopp@:~/pro/git/otclient/build $ cmake .. -USE_PROTOBUF -DVORBISFILE_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libvorbisfile.so -DVORBIS_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libvorbis.so
...

scopp@:~/pro/git/otclient/build $ make -j13
[  0%] Building CXX object CMakeFiles/otclient.dir/src/framework/luafunctions.cpp.o
[  1%] Building CXX object CMakeFiles/otclient.dir/src/framework/util/color.cpp.o
[  2%] Building CXX object CMakeFiles/otclient.dir/src/framework/util/crypt.cpp.o

... continues without any error, and without compiling protobuf files
```

**Example building executing cmake twice**
```sh
scopp@:~/pro/git/otclient/build $ rm ../src/client/protobuf/*.pb.*
scopp@:~/pro/git/otclient/build $ rm -fr *
scopp@:~/pro/git/otclient/build $ cmake .. -USE_PROTOBUF -DVORBISFILE_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libvorbisfile.so -DVORBIS_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libvorbis.so
...
scopp@:~/pro/git/otclient/build $ cmake .. -USE_PROTOBUF -DVORBISFILE_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libvorbisfile.so -DVORBIS_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libvorbis.so
...

scopp@:~/pro/git/otclient/build $ make -j13
[  1%] Compiling shared.proto 		<-- as it should be, because I deleted the compiled files in the first instruction
[  1%] Compiling appearances.proto  <-- as it should be, because I deleted the compiled files in the first instruction
[  1%] Building CXX object CMakeFiles/otclient.dir/src/framework/luafunctions.cpp.o
[  2%] Building CXX object CMakeFiles/otclient.dir/src/framework/util/color.cpp.o
[  3%] Building CXX object CMakeFiles/otclient.dir/src/framework/util/crypt.cpp.o

... continues, and when finishing:

[ 98%] Building CXX object CMakeFiles/otclient.dir/src/client/uisprite.cpp.o
[ 99%] Building CXX object CMakeFiles/otclient.dir/src/main.cpp.o
[100%] Linking CXX executable otclient
/usr/bin/ld: CMakeFiles/otclient.dir/src/client/protobuf/appearances.pb.cc.o: in function `tibia::protobuf::appearances::SpritePhase::ByteSizeLong() const':

... and a big log of errors
```

**Comment by nekiro** [comment]
If you guys are not in a hurry, I have a plan to implement this dll https://github.com/nekiro/ProtobufLib into this repository.
Then it should be way easier to require or not require protobuf.

**Comment by tornadia** [comment]
A bit busy myself irl. I will check the issues provided by @scopz and changes planned by @nekiro soon.

**Pull request closed without merging**